impl/src/main/java/com/okta/sdk/impl/http/authc/OAuth2RequestAuthenticator.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2020-Present Okta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.okta.sdk.impl.http.authc;
+
+import com.okta.commons.http.Request;
+import com.okta.commons.http.authc.RequestAuthenticationException;
+import com.okta.commons.http.authc.RequestAuthenticator;
+import com.okta.commons.lang.Assert;
+import com.okta.sdk.authc.credentials.ClientCredentials;
+import com.okta.sdk.impl.oauth2.OAuth2AccessToken;
+import com.okta.sdk.impl.oauth2.OAuth2ClientCredentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.security.InvalidKeyException;
+
+/**
+ * This implementation used by OAuth2 flow adds Bearer header with access token as the
+ * value in all outgoing requests. This has logic to fetch a new token and store it in
+ * {@link OAuth2ClientCredentials} object if the existing one has expired.
+ *
+ * @since 1.6.0
+ */
+public class OAuth2RequestAuthenticator implements RequestAuthenticator {
+    private static final Logger log = LoggerFactory.getLogger(OAuth2RequestAuthenticator.class);
+
+    private final ClientCredentials<OAuth2AccessToken> clientCredentials;
+
+    public OAuth2RequestAuthenticator(ClientCredentials<OAuth2AccessToken> clientCredentials) {
+        Assert.notNull(clientCredentials, "clientCredentials may not be null");
+        this.clientCredentials = clientCredentials;
+    }
+
+    @Override
+    public void authenticate(Request request) throws RequestAuthenticationException {
+        OAuth2AccessToken oAuth2AccessToken = clientCredentials.getCredentials();
+
+        if (oAuth2AccessToken.hasExpired()) {
```

Comment:
`oAuth2AccessToken` can and will be null here, (it's explicitly set to `null` below when `reset()` is called).
This case needs to be handled, maybe adding a null check to the if statements 🤔

Reply:
(i.e. when line 60 is called then line 50 is called before line 64)

Reply:
I've added code to retrieve access token with it is null here. 

However, I cannot add a null check (double null check for thread safety reasons) in later lines because of a PMD bug - https://github.com/spotbugs/spotbugs/issues/651. 

We could still add a check and suppress the PMD violation by doing something like https://stackoverflow.com/questions/1829904/is-there-a-way-to-ignore-a-single-findbugs-warning. However, this needs an additional dependency on:

`<dependency>
      <groupId>com.google.code.findbugs</groupId>
      <artifactId>annotations</artifactId>
      <version>3.0.0</version>
      <scope>provided</scope>
</dependency>`

Are you fine with this approach?

Reply:
As discussed on Slack, I've added double null checking for thread safety and excluded this rule in findbugs.

Reply:
Refactored code around the concurrency problems we discussed on Zoom today. Also, wrote some unit tests around this and ensured code coverage.